src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php

```diff
@@ -0,0 +1,237 @@
+            $allowedExtraValues[] = $defaultValue;
+        }
+
+        foreach ($allowedExtraValues as $i => $allowedExtraValue) {
+            if (null !== $allowedExtraValue && !\is_scalar($allowedExtraValue)) {
+                // IDEs don't seem to understand non-scalar values in "enum", so let's create separate sub-schema
+                unset($allowedExtraValues[$i]);
+                if (\is_array($allowedExtraValue) && array_is_list($allowedExtraValue)) {
+                    $subSchemas[] = (object) ['const' => $allowedExtraValue];
+                }
+            }
+        }
+
+        if ($allowedValues = array_values(array_filter($allowedExtraValues, static fn (mixed $value) => !$validateValue($value)))) {
+            if (!isset($schema->enum)) {
+                // Append "boolean" to "type" instead of [true, false] to "enum"
+                if (false !== ($true = array_search(true, $allowedValues, true)) && false !== ($false = array_search(false, $allowedValues, true))) {
+                    unset($allowedValues[$true], $allowedValues[$false]);
+                    $this->addTypeToSchema($schema, 'boolean');
+                }
+
+                // Append "null" to "type" instead of null to "enum"
+                if (false !== ($null = array_search(null, $allowedValues, true))) {
+                    unset($allowedValues[$null]);
+                    $this->addTypeToSchema($schema, 'null');
+                }
+
+                if ($allowedValues) {
+                    $subSchemas[] = (object) ['enum' => array_values($allowedValues)];
+                }
+            } else {
+                $schema->enum = array_values(array_unique(array_merge($schema->enum, $allowedValues)));
+            }
+        }
+
+        if ($schema) {
+            $subSchemas[] = $schema;
+            if ($allowParam && !\in_array('string', (array) ($schema->type ?? null))) {
+                $subSchemas[] = (object) ['$ref' => '#/definitions/param'];
+            }
+        }
+
+        $subSchemas ??= [new \stdClass()];
+
+        return $subSchemas;
+    }
+
+    private function createArraySchema(PrototypedArrayNode $node, \stdClass $definitions): \stdClass
+    {
+        $prototypeSchema = $this->buildSingleNode($node->getPrototype(), $definitions);
+        $prototypeRef = $this->getReference($prototypeSchema, $definitions);
+
+        $schema = (object) [
+            'type' => ['array', 'object'],
+            'items' => $prototypeRef,
+            'additionalProperties' => $prototypeRef,
+        ];
+
+        if ($node->getMinNumberOfElements() > 0) {
+            $schema->minItems = $schema->minProperties = $node->getMinNumberOfElements();
+        }
+
+        return $schema;
+    }
+
+    private function createObjectSchema(ArrayNode $node, \stdClass $definitions): \stdClass
+    {
+        $schema = (object) [
+            'type' => 'object',
```

Comment:
If additional properties are allowed, we loose the validation on property names.
I think they can be disallowed once alternative documentation formats are fully described: https://github.com/symfony/symfony/pull/51273
```suggestion
            'type' => 'object',
            'additionalProperties' => false,
```